AzureIoTHub.agent.lib.nut

```diff
@@ -629,513 +632,819 @@ class AzureIoTHub {
+            } else {
+                try {
+                    // We are connected or connecting now
+                    _onConnectCb(_state & AZURE_IOT_CLIENT_STATE_CONNECTED ? AZURE_IOT_CLIENT_ERROR_ALREADY_CONNECTED : AZURE_IOT_CLIENT_ERROR_OP_NOT_ALLOWED_NOW);
+                } catch (e) {
+                    _error("Exception at onConnect user callback: " + e);
+                }
+            }
         }
 
-        function _openEventSender() {
-            _senders.event <- _sessions.event.opensender(AzureIoTHub.Endpoint.eventPath(_config.deviceId), _amqpEventSenderStatusHandler.bindenv(this));
+        // Closes the connection to Azure IoT Hub. Does nothing if the connection is already closed.
+        //
+        // Returns:                         Nothing.
+        function disconnect() {
+            if (_state != AZURE_IOT_CLIENT_STATE_DISCONNECTED && !(_state & AZURE_IOT_CLIENT_STATE_DISCONNECTING)) {
+                _state = _state | AZURE_IOT_CLIENT_STATE_DISCONNECTING;
+                _mqttclient.disconnect(_onDisconnect.bindenv(this));
+            }
         }
 
-        function _openEventReceiver() {
-            _receivers.event <- _sessions.event.openreceiver(AzureIoTHub.Endpoint.messagePath(_config.deviceId), _amqpEventReceiverStatusHandler.bindenv(this), function(deliveries) {
-                _handleDeliveries(deliveries, _handlers.onEvent);
-            }.bindenv(this));
-        }
+        // Checks if the client is connected to Azure IoT Hub.
+        //
+        // Returns:                         Boolean: true if the client is connected, false otherwise.
+        function isConnected() {
+            return _state & AZURE_IOT_CLIENT_STATE_CONNECTED > 0;
+        }
+
+        // Sends a message to Azure IoT Hub (https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-mqtt-support#sending-device-to-cloud-messages).
+        //
+        // Parameters:
+        //     msg : AzureIoTHub.Message    Message to send.
+        //     onComplete : Function        Callback called when the operation is completed or an error happens.
+        //          (optional)              The callback signature:
+        //                                  onComplete(error), where
+        //                                      error : Integer     0 if the operation is completed successfully, an error code otherwise.
+        //
+        // Returns:                         Nothing.
+        function sendMessage(msg, onComplete = null) {
+            if ((_state & AZURE_IOT_CLIENT_STATE_CONNECTED) && !(_state & AZURE_IOT_CLIENT_STATE_DISCONNECTING)) {
+                if (_msgSentCbMap.len() >= _options.sendMsgParallelReqs) {
+                    _callOnComplete(onComplete, AZURE_IOT_CLIENT_ERROR_OP_NOT_ALLOWED_NOW);
+                    return;
+                }
+                // TODO: http.urlencode follows RFC3986, but Azure expects RFC2396 string. Should we worry about it?
```

Comment:
As far as I see, the primary difference between the specs is the URL resolution. while we only encode parameters, which is just escaping I assume. or am I missing something?

Reply:
I think so too.
Done